fix multiple-fake-modes bug with compile + subclasses

bdhirsh opened the pull request:
This should fix the "multiple fake modes" errors we've been seeing with both float8 tensor and DTensor.

Haven't added a test yet - will add one before landing.

I also have a separate PR that would have made the error significantly nicer (the bad error resulted from us returning a FakeTensor at runtime): https://github.com/pytorch/pytorch/pull/118644

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119948
* #119067
* #118802
* #118671
* #118670
* #118669
* #119947
* #118803
* #118668
* #118667
* __->__ #118191
* #118209
* #117666
* #117667



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

bdhirsh committed:
[test] fix multiple-fake-modes bug with compile + subclasses

[ghstack-poisoned]

ezyang commented:
I guess you'll add a test if this worked? I'd love to see the analysis that led to this changed.

bdhirsh committed:
Update on "[test] fix multiple-fake-modes bug with compile + subclasses"




[ghstack-poisoned]

bdhirsh commented:
Analysis / issue filed here: https://github.com/pytorch/pytorch/issues/118799. Will add a test soon (the one in the issue)

bdhirsh committed:
Update on "fix multiple-fake-modes bug with compile + subclasses"

This should fix the "multiple fake modes" errors we've been seeing with both float8 tensor and DTensor.

Haven't added a test yet - will add one before landing.

I also have a separate PR that would have made the error significantly nicer (the bad error resulted from us returning a FakeTensor at runtime): https://github.com/pytorch/pytorch/pull/118644




[ghstack-poisoned]

bdhirsh committed:
Update on "fix multiple-fake-modes bug with compile + subclasses"

This should fix the "multiple fake modes" errors we've been seeing with both float8 tensor and DTensor.

Haven't added a test yet - will add one before landing.

I also have a separate PR that would have made the error significantly nicer (the bad error resulted from us returning a FakeTensor at runtime): https://github.com/pytorch/pytorch/pull/118644




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

bdhirsh committed:
Update on "fix multiple-fake-modes bug with compile + subclasses"

This should fix the "multiple fake modes" errors we've been seeing with both float8 tensor and DTensor.

Haven't added a test yet - will add one before landing.

I also have a separate PR that would have made the error significantly nicer (the bad error resulted from us returning a FakeTensor at runtime): https://github.com/pytorch/pytorch/pull/118644




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

bdhirsh committed:
Update on "fix multiple-fake-modes bug with compile + subclasses"

This should fix the "multiple fake modes" errors we've been seeing with both float8 tensor and DTensor.

Haven't added a test yet - will add one before landing.

I also have a separate PR that would have made the error significantly nicer (the bad error resulted from us returning a FakeTensor at runtime): https://github.com/pytorch/pytorch/pull/118644




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

drisspg commented on torch/_functorch/_aot_autograd/subclass_utils.py:
```diff
@@ -24,10 +24,10 @@ def requires_subclass_dispatch(args, fw_metadata: ViewAndMutationMeta) -> bool:
-        is_traceable_wrapper_subclass(x)
-        for x in fw_metadata.traced_tangents
-        if isinstance(x, Tensor)
+        type(x) is SubclassCreationMeta for x in fw_metadata.subclass_fw_graph_out_meta
```
Noob q; 
From reading the issue you linked,  I take it that the problem wasn't really the  "check" for subclass outputs but really that we weren't checking enough things?

For the check on inputs is that exhaustive enough?

bdhirsh replied:
> I take it that the problem wasn't really the "check" for subclass outputs but really that we weren't checking enough things?

Yep this is right

> For the check on inputs is that exhaustive enough?

yeah - the reason the check wasn't exhaustive in the first place is because "graph outputs" and "tangents" can be different (there can be extra outputs to the forward graph, that don't get tangents). But "graph inputs" doesn't have any extra complexity to it - we just loop through them all right now
